File: backend/tests/taxi_pricing.test.js
const { calculateFare, fareForType } = require('../services/taxi_pricing_service');

describe('Taxi Pricing Service', () => {
  describe('fareForType', () => {
    it('tuktuk up to 2km is 1000', () => {
      expect(fareForType(1.0, 'tuktuk')).toBe(1000);
      expect(fareForType(2.0, 'tuktuk')).toBe(1000);
    });

    it('tuktuk 3km is 1250', () => {
      expect(fareForType(3.0, 'tuktuk')).toBe(1250);
    });

    it('wazz up to 2km is 1500', () => {
      expect(fareForType(1.5, 'wazz')).toBe(1500);
      expect(fareForType(2.0, 'wazz')).toBe(1500);
    });

    it('wazz 3km is 1800', () => {
      expect(fareForType(3.0, 'wazz')).toBe(1800);
    });

    it('economic minimum 1500 for 1km', () => {
      expect(fareForType(1.0, 'economic')).toBe(1500);
      expect(fareForType(2.0, 'economic')).toBe(1500);
    });

    it('economic 3km is 2000', () => {
      expect(fareForType(3.0, 'economic')).toBe(2000);
    });

    it('economic 5km is 3000', () => {
      expect(fareForType(5.0, 'economic')).toBe(3000);
    });

    it('rounds to nearest 250 IQD', () => {
      const { roundFareToNearestStep } = require('../services/taxi_pricing_service');
      expect(roundFareToNearestStep(1430)).toBe(1500);
      expect(roundFareToNearestStep(1700)).toBe(1700);
      expect(roundFareToNearestStep(1370)).toBe(1250);
    });
      expect(fareForType(1.0, 'super')).toBe(1500);
    });

    it('caps at 50000', () => {
      expect(fareForType(200, 'economic')).toBe(50000);
    });
  });

  describe('calculateFare', () => {
    it('returns fare for selected type', () => {
      const result = calculateFare(3.0, 'tuktuk');
      expect(result.fare).toBe(1250);
      expect(result.fareEconomic).toBe(2000);
    });
  });
});

File: test/taxi_pricing_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:alghaith_app/features/taxi/utils/taxi_fare_calculator.dart';
import 'package:alghaith_app/features/taxi/models/taxi_request.dart';

void main() {
  group('TaxiFareCalculator', () {
    test('tuktuk: up to 2km is 1000', () {
      expect(TaxiFareCalculator.fareForType(1.0, TaxiType.tuktuk), 1000);
      expect(TaxiFareCalculator.fareForType(2.0, TaxiType.tuktuk), 1000);
    });

    test('tuktuk: 3km is 1250', () {
      expect(TaxiFareCalculator.fareForType(3.0, TaxiType.tuktuk), 1250);
    });

    test('wazz: up to 2km is 1500', () {
      expect(TaxiFareCalculator.fareForType(1.0, TaxiType.wazz), 1500);
      expect(TaxiFareCalculator.fareForType(2.0, TaxiType.wazz), 1500);
    });

    test('wazz: 3km is 1800', () {
      expect(TaxiFareCalculator.fareForType(3.0, TaxiType.wazz), 1800);
    });

    test('economic: minimum 1500 even for 1km', () {
      expect(TaxiFareCalculator.fareForType(1.0, TaxiType.economic), 1500);
      expect(TaxiFareCalculator.fareForType(2.0, TaxiType.economic), 1500);
    });

    test('economic: 3km is 2000', () {
      expect(TaxiFareCalculator.fareForType(3.0, TaxiType.economic), 2000);
    });

    test('economic: 5km is 3000', () {
      expect(TaxiFareCalculator.fareForType(5.0, TaxiType.economic), 3000);
    });

    test('caps at 50000 for long distance', () {
      expect(TaxiFareCalculator.fareForType(200.0, TaxiType.economic), 50000);
    });

    test('rounds to nearest 250 IQD', () {
      expect(TaxiFareCalculator.roundFareToNearestStep(1430), 1500);
      expect(TaxiFareCalculator.roundFareToNearestStep(1700), 1700);
      expect(TaxiFareCalculator.roundFareToNearestStep(1370), 1250);
    });

    test('TaxiType API mapping', () {
      expect(TaxiTypeX.fromApiName('tuktuk'), TaxiType.tuktuk);
      expect(TaxiTypeX.fromApiName('wazz'), TaxiType.wazz);
      expect(TaxiTypeX.fromApiName('super'), TaxiType.economic);
      expect(TaxiType.tuktuk.toApiName, 'tuktuk');
    });
  });
}
